### ISSUE TITLE
use `scala-cli fmt` formatting when scala-cli BSP is used

### ISSUE THREAD

**Issue by danielleontiev** [scalameta/metals-feature-requests#334]
### Is your feature request related to a problem? Please describe.

When writing scala-cli scripts it's common to not set up the whole environment (f.e. adding .scalafmt.conf near the script) but it's desirable to be able to format the code using the same metals formatting command that is used for "real" projects that have .scalafmt.conf in the root directory.

### Describe the solution you'd like

To address the above example scala-cli has embedded scalafmt with default configuration that does not require creating .scalafmt.conf (it can be issued with `scala-cli fmt` command) - it may be beneficial to run scala-cli's embedded formatting command in such cases from metals

### Describe alternatives you've considered

Another option would be to just have .scalafmt.conf fallback for metals since metals have embedded scalafmt as well

### Search terms

scala-cli fmt formatting scalafmt

**Comment by adpi2** [comment]
Thanks for opening this feature request.

In general I think we should avoid special-casing on the build tools, and we can do so using BSP. So the correct way to implement this feature would be to add a `format` request in BSP and to implement it in scala-cli. There should be a `format` capability as well so that Metals knows if the build tool supports formatting or not. If it supports it, Metals sends a BSP `format` request, otherwise it calls `scalafmt` itself.

**Comment by danielleontiev** [comment]
Yes, that's very helpful, thank you! I will try to investigate why scala-cli does not implement `format` BSP call

**Comment by danielleontiev** [comment]
As far as I can see BSP specification does not have `format` request: https://build-server-protocol.github.io/docs/specification/

Do you mean that BSP protocol shold be extended to support this?

**Comment by adpi2** [comment]
> Do you mean that BSP protocol shold be extended to support this?

Yes, I think we should extend it with a `buildTarget/format` request and a `canFormat` capability (in the [`BuildTargetCapabilities`](https://build-server-protocol.github.io/docs/specification/#build-target)).

**Comment by danielleontiev** [comment]
Thinking about current BSP implementations for Scala such as `sbt`, `mill`, `bloop` or `scala-cli` I do not know whether such capability would fit to any BSP other than `scala-cli` (just because `scala-cli` being a universal Swiss knife) - won't we end up in the situation when the whole protocol method exist only to satisfy one concrete implementation (`scala-cli`)? Will it be beneficial to extend protocol and add `buildTarget/format` in comparison with straightforward solution like adding default `.scalafmt.conf` for Metals? I am asking because I am a bit confused about responsibilities that BSP and LSP should handle - thinking about code formatting does not seem to be a clear BSP responsibility (from my pure user-side point of view).

If I am wrong and we do need to extend the protocol - what should be the next steps? Does opening a PR with new method and capability in https://github.com/build-server-protocol/build-server-protocol will work?

**Comment by adpi2** [comment]
> won't we end up in the situation when the whole protocol method exist only to satisfy one concrete implementation (scala-cli)? Will it be beneficial to extend protocol and add buildTarget/format in comparison with straightforward solution like adding default .scalafmt.conf for Metals?

Maybe you are right and it would be easier to implement a straightforward solution in Metals. But what about IntelliJ then? Will you implement the special case handling in IntelliJ as well?

> I do not know whether such capability would fit to any BSP other than scala-cli

As long as we have a `.scalafmt.conf` file, we indeed don't need to implement this capability in sbt, Mill or Bloop because Metals or IntelliJ can invoke scalafmt in a reproducible and consistent way. It would be possible though to implement it, for instance in sbt, in the `sbt-scalafmt` plugin. If one adds the `sbt-scalafmt` plugin then the `canFormat` capability becomes true.

In general, I thing the `buildTarget/format` request makes sense because you can have more than just Scala code in your build, maybe you have Java code, or html and css files, or Twirl files. Should Metals know how to format all of them? I rather think it is the responsibility of the build tool to format all those different languages.

In conclusion I think it would be worth defining the `buildTarget/format` request in BSP and I would encourage you to do so. If you prefer to implement the quick fix in Metals directly, that's fine too.

**Comment by danielleontiev** [comment]
Thank you a lot for comprehensive explanation, I think I am convinced now about `buildTarget/format`.

> I would encourage you to do so

How can I approach this? Is it ok to create PR in this repository https://github.com/build-server-protocol/build-server-protocol or some prerequisites are necessary?

**Comment by adpi2** [comment]
> Is it ok to create PR in this repository https://github.com/build-server-protocol/build-server-protocol or some prerequisites are necessary

Yes, that's the way to go!

**Comment by danielleontiev** [comment]
Thank you, I will do it soon

**Comment by danielleontiev** [comment]
I will close this one for now because it's decided to not taking BSP formatting approach for this. But if becomes relevant it may be reopened with additional details
